lib/double_entry/reporting/line_metadata_filter.rb

```diff
@@ -0,0 +1,33 @@
+# encoding: utf-8
+module DoubleEntry
+  module Reporting
+    class LineMetadataFilter
+
+      def self.filter(collection:, metadata:)
+        table_alias_index = 0
+
+        metadata.reduce(collection) do |filtered_collection, (key, value)|
+          table_alias = "m#{table_alias_index}"
+          table_alias_index += 1
+
+          filtered_collection.
+            joins("INNER JOIN #{line_metadata_table} as #{table_alias} ON #{table_alias}.line_id = #{lines_table}.id").
```

Comment:
I posed two approaches to the query to the team yesterday, we decided on this one (the other was more convoluted and involved grouping by the line ID & checking the number of occurrences, but it did use a single join).

Basically, we want to make sure that the line has all of these metadata key/pair values.  We can't just do something like `(WHERE key = 'key1' AND value = 'value1') AND (WHERE key = 'key2' AND value = 'value2')` because key cannot be both key1 and key2 at the same time.  And once we've filtered out the `line_metadata` results by one key, we can't do the same with the second using the same table we've joined.  So we need to join `line_metadata` again to get the other key.